packages/main/src/plugin/proxy-resolver.ts

```diff
@@ -0,0 +1,135 @@
+  if (!proxyUrl) return;
+  const options = {
+    keepAlive: true,
+    keepAliveMsecs: 1000,
+    maxSockets: 256,
+    maxFreeSockets: 256,
+    scheduling: 'lifo',
+    proxy: proxyUrl,
+  };
+  return secure
+    ? new HttpsProxyAgent(options as HttpProxyAgentOptions)
+    : new HttpProxyAgent(options as HttpsProxyAgentOptions);
+}
+
+function getProxyUrl(proxy: Proxy): string | undefined {
+  if (proxy.isEnabled()) {
+    return proxy.proxy?.httpsProxy ? proxy.proxy?.httpsProxy : proxy.proxy?.httpProxy;
+  }
+}
+
+export function getOptions(proxy: Proxy, secure: boolean): { agent?: http.Agent | https.Agent } {
+  const agent = createProxyAgent(secure, getProxyUrl(proxy));
+  return agent ? { agent } : {};
+}
+
+export function createHttpPatch(originals: typeof http | typeof https, proxy: Proxy) {
+  return {
+    get: patch(originals.get),
+    request: patch(originals.request),
+  };
+
+  function patch(original: typeof http.get) {
+    function patched(
+      url?: string | nodeurl.URL | null,
+      options?: http.RequestOptions | null,
+      callback?: (res: http.IncomingMessage) => void,
+    ): http.ClientRequest {
+      if (typeof url !== 'string' && !url?.searchParams) {
+        callback = <any>options; // eslint-disable-line @typescript-eslint/no-explicit-any
+        options = url;
+        url = null;
+      }
+      if (typeof options === 'function') {
+        callback = options;
+        options = null;
+      }
+      options = options || {};
+
+      if (options.socketPath) {
+        return original(options, callback);
+      }
+
+      const originalAgent = options.agent;
+      if (originalAgent === true) {
+        throw new Error('Unexpected agent option: true');
+      }
+
+      if (proxy?.isEnabled()) {
+        if (url) {
+          const parsed = typeof url === 'string' ? new nodeurl.URL(url) : url;
+          const urlOptions = {
+            protocol: parsed.protocol,
+            hostname: parsed.hostname.lastIndexOf('[', 0) === 0 ? parsed.hostname.slice(1, -1) : parsed.hostname,
+            port: parsed.port,
+            path: `${parsed.pathname}${parsed.search}`,
+          };
+          if (parsed.username || parsed.password) {
+            options.auth = `${parsed.username}:${parsed.password}`;
+          }
+          options = { ...urlOptions, ...options };
+        } else {
+          options = { ...options };
+        }
+
+        const host = options.hostname || options.host;
+        const isLocalhost = !host || host === 'localhost' || host === '127.0.0.1'; // Avoiding https://github.com/microsoft/vscode/issues/120354
+        if (!isLocalhost) {
+          options = Object.assign({}, options, getOptions(proxy, options.protocol === 'https'));
+        }
+
+        return original(options, callback);
+      }
+      return original.apply(null, arguments as any); // eslint-disable-line
+    }
+    return patched;
+  }
+}
+
+export function createPatchedModules(proxy: Proxy) {
```

Comment:
WDYT about renaming it createHttpPatchedModules so from consumer pov we know what it'll do.